KEP-5721: Add Semantic Version Comparison Operators to Tolerations and NodeAffinity

galal-hussein opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: Add Semantic Version Comparison Operators to Tolerations and NodeAffinity

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/5721

<!-- other comments or additional information -->
- Other comments:

galal-hussein committed:
KEP-5721 Semver Operators

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
Add persistent volume part in the KEP

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
fix design implementation

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
update toc

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
Add DRA nodeSelector utilization of semver and update unit tests

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
Fix kep.yaml

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
Update TOC

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein commented:
/retest

helayoty commented:
/assign

galal-hussein committed:
fixes

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

helayoty commented on keps/sig-scheduling/5721-semver-operators/kep.yaml:
```diff
@@ -0,0 +1,41 @@
+reviewers:
+  - TBD
+approvers:
+  - TBD
```
Add @macsko and @sanposhiho as approvers

helayoty commented on keps/sig-scheduling/5721-semver-operators/README.md:
```diff
@@ -0,0 +1,1393 @@
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
```
Please update the PRR checklist

helayoty commented on keps/sig-scheduling/5721-semver-operators/README.md:
````diff
@@ -0,0 +1,1393 @@
+            operator: "SemverGt"
+            values: ["2.0.0"]
+```
+#### Story 4 — Persistent Volume node affinity based on kernel version
````
```suggestion

#### Story 4 — Persistent Volume node affinity based on kernel version
```

helayoty commented on keps/sig-scheduling/5721-semver-operators/README.md:
````diff
@@ -0,0 +1,1393 @@
+      allNodes: true
+```
+
+This ensures that GPU devices in this ResourceSlice are only considered for allocation when the node meets the version requirements, enabling workloads that require newer CUDA features or kernel capabilities to be scheduled appropriately.
````
We can remove this as extra details for the user story section.

helayoty commented on keps/sig-scheduling/5721-semver-operators/README.md:
```diff
@@ -0,0 +1,1393 @@
+- `SemverLt`: match if version of value of NodeSelectorRequirement.Key < version of NodeSelectorRequirement.Values[0]
+- `SemverGt`: match if version of value of NodeSelectorRequirement.Key > version of NodeSelectorRequirement.Values[0]
+- `SemverEq`: match if version of value of NodeSelectorRequirement.Key = version of NodeSelectorRequirement.Values[0]
+- `In`/`NotIn`/`Exists`/`DoesNotExists`/`Gt`/`Lt`: Remain unchanged
```
```suggestion
```

helayoty commented on keps/sig-scheduling/5721-semver-operators/README.md:
````diff
@@ -0,0 +1,1393 @@
+  NodeSelectorOpSemverEq  NodeSelectorOperator = "SemverEq"
+)
+
+```
````
```suggestion
```go
type NodeSelectorOperator string

const (
	...
  // New semver comparison operators
  NodeSelectorOpSemverGt  NodeSelectorOperator = "SemverGt"
  NodeSelectorOpSemverLt  NodeSelectorOperator = "SemverLt"
  NodeSelectorOpSemverEq  NodeSelectorOperator = "SemverEq"
)

```
```

helayoty commented on keps/sig-scheduling/5721-semver-operators/README.md:
```diff
@@ -0,0 +1,1393 @@
+
+###### How can someone using this feature know that it is working for their instance?
+
+<!--
```
Please remove all KEP template comments to make it easier for the reviewers.

helayoty reviewed (commented):
First round of review.

galal-hussein committed:
fixes to Readme and adding prr yaml file

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
update kep.yaml file with reviewers

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
remove metrics part from kep.yaml

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
update production readiness file

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
fix logic

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein commented on keps/sig-scheduling/5721-semver-operators/kep.yaml:
```diff
@@ -0,0 +1,41 @@
+# The following PRR answers are required at alpha release
+# List the feature gate name and the components for which it must be enabled
+feature-gates:
+  - name: TaintTolerationNodeAffinitySemverComparisonOperatorss
```
I have changed it to `TolerationAffinitySemverOperators` if thats okay, to include the toleration and affinity

galal-hussein committed:
remove file paths in design details

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
fix checklist

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein commented:
> First round of review.

@helayoty Thanks for the review, I have updated the PR addressing your comments, please let me know if the changes are okay.

helayoty commented on keps/sig-scheduling/5721-semver-operators/README.md:
```diff
@@ -0,0 +1,734 @@
+- **Node Affinity**: If a node label value cannot be parsed as SemVer, the affinity matching returns `false`:
+  - For `requiredDuringSchedulingIgnoredDuringExecution`: Pod cannot schedule on that node
+  - For `preferredDuringSchedulingIgnoredDuringExecution`: Node receives 0 score contribution for that term (pod can still schedule)
+  - Scheduler logs (V(10) level): `"Parse semver failed for value X in label Y"`
```
```suggestion
```

helayoty commented on keps/sig-scheduling/5721-semver-operators/README.md:
````diff
@@ -0,0 +1,734 @@
+  storageClassName: advanced-storage
+```
+
+### Notes/Constraints/Caveats (Optional)
````
```suggestion
### Notes/Constraints/Caveats
```

galal-hussein committed:
update alternative section and add more info for semver library

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
remove optional remark

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
fix typo

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

dom4ha commented on keps/sig-scheduling/5721-semver-operators/README.md:
```diff
@@ -0,0 +1,779 @@
+
+- **Toleration Parsing Requirements**: The toleration value must be parseable as Semantic Versions for SemVer operators (`SemverLt`, `SemverGt`, `SemverEq`). If parsing fails, the toleration does not match.
+
+- **Node Affinity Parsing Requirements**: The node affinity value must be parseable as Semantic Versions for SemVer operators (`SemverLt`, `SemverGt`, `SemverEq`). If parsing fails, the node affinity matching rule does not match.
```
When parsing fails, is it going to be reflected in Unschedulable status?

soltysh reviewed (commented):
Some comments from PRR review, which seems to align with concerns from sig-scheduling leads.
